docs/source/api/notebooks/Notebook_5_GP_solution.ipynb

```diff
@@ -0,0 +1,631 @@
+    "        base_encoder: GraphStorm ConvEncoder types, options:\n",
+    "            1. `RelationalGCNEncoder`;\n",
+    "            2. `RelationalGATEncoder`;\n",
+    "            3. `HGTEncoder`.\n",
+    "        read_out_opt: string in the following options:\n",
+    "            The aggregation method for the cached supernodes' representations. The current options:\n",
+    "            1. `last_only`: only use the last layer's representations. If use this option, \n",
+    "               the read_out_ops will be ignored because there is only one layer representation\n",
+    "               is involved in the final read_out.\n",
+    "            2. `mean`: compute the mean of all of the cached supernode representations.\n",
+    "            3. `sum`: compute the summantion of all of the cached supernode representations.\n",
+    "            4. `weighted_sum`: use additional weight parameters to compute the weighted summation\n",
+    "               all of the cached supernode representations.\n",
+    "            5. `min`: compute the minimum in each dimension of the all of the cached supernode\n",
+    "               representations.\n",
+    "            6. `max`: compute the maximum in each dimension of the all of the cached supernode\n",
+    "               representations.\n",
+    "        super_ntype: string\n",
+    "            The name of supernode type. Default is 'super'.\n",
+    "\n",
+    "    \"\"\"\n",
+    "    def __init__(self,\n",
+    "                 h_dim,\n",
+    "                 out_dim,\n",
+    "                 base_encoder,\n",
+    "                 read_out_opt='last_only',\n",
+    "                 super_ntype='super'\n",
+    "                ):\n",
+    "        assert isinstance(base_encoder, (RelationalGCNEncoder, RelationalGATEncoder, HGTEncoder)), \\\n",
+    "               'Only support GraphStorm\\'s RelationalGCNEncoder, RelationalGATEncoder, and HGTEncoder'\n",
+    "        assert base_encoder.num_layers >= 3, 'For Graph Prediction task, at least two layers GNN' + \\\n",
+    "                                       f'encoder required, but got {num_hidden_layers} ...'\n",
+    "        super(GPEncoder4SupernodeOgbg, self).__init__(h_dim, out_dim, base_encoder.num_layers)\n",
+    "\n",
+    "        assert read_out_opt in ['last_only', 'mean', 'sum', 'weighted_sum', 'min', 'max'], + \\\n",
+    "                                f'Not recognized read_out_opt {read_out_opt}. Options include ' + \\\n",
+    "                                '\\'last_only\\', \\'mean\\', \\'sum\\', \\'weighted_sum\\', \\'mim\\', ' + \\\n",
+    "                                'and \\'max\\'.'\n",
+    "        self.base_encoder = base_encoder\n",
+    "        self.read_out_opt = read_out_opt\n",
+    "        self.super_ntype = super_ntype\n",
+    "        if read_out_opt=='weighted_sum':\n",
+    "            self.weighted_sum_para = th.nn.Parameter(th.Tensor(1, num_hidden_layers))\n",
```

Comment:
Instead of defining a class attribute conditionally it's usually safer to assign to None or the intended value.

Otherwise code like

```
if self.weighted_sum_para:
   pass
```

will trigger an AttributeError

Reply:
As this is just an example, I will set this parameter to None by default.